sympy/sets/sets.py

```diff
@@ -652,6 +707,19 @@ def _boundary(self):
     def _measure(self):
         raise NotImplementedError("(%s)._measure" % self)
 
+    @property
+    def _kind(self):
+        if not self.args:
+            return SetKind()
+        elif all(i.kind == self.args[0].kind for i in self.args):
+            return SetKind(self.args[0].kind)
+        else:
+            return SetKind(UndefinedKind)
```

Comment:
It's a bad idea to make these presumptions about the meaning of `args` in a superclass like `Set`. This method should probably just return `SetKind(UndefinedKind)`. It's up to subclasses to be more specific.

Reply:
I added this logic here because this method is used in almost every subclasses, if we remove this, same method has to be used 3-4 times in subclasses.
Even if presumptions about args fails, it will go to default case of `SetKind(UndefinedKind)`. So is there a need to change it?

Reply:
We can add a helper method to implement this logic (as I suggested above) but it shouldn't be called by default in the `Set` superclass. Each class should explicitly call the helper if appropriate.

Reply:
if we just return `SetKind(UndefinedKind)` in set class, then kind of PowerSet will only be `SetKind(UndefinedKind)`

Reply:
The PowerSet method should override this. It's not hard to implement the PowerSet.kind method:
```python
def kind(self):
    return SetKind(self.arg.kind)
```

Reply:
```python
    @property
    def _kind(self):
        return SetKind(UndefinedKind)

    @property
    def _helper_kind(self):
        if not self.args:
            return SetKind()
        elif all(i.kind == self.args[0].kind for i in self.args):
            return SetKind(self.args[0].kind)
        else:
            return SetKind(UndefinedKind)
    
    @property
    def _union_setkind(self):
        args = tuple(arg.kind for arg in self.args if arg is not S.EmptySet)
        if not args:
            return SetKind()
        elif all(i == args[0] for i in args):
            return args[0]
        else:
            return SetKind(UndefinedKind)

    @property
    def _intersection_setkind(self):
        args = tuple(arg.kind for arg in self.args if arg is not S.UniversalSet)
        if not args:
            return SetKind(UndefinedKind)
        elif all(i == args[0] for i in args):
            return args[0]
        else:
            return SetKind()
```
Will this work?

Reply:
I don't see why any of these are property methods. It would make more sense for a helper to be `classmethod`. 

The intersection method only seems to be used by the intersection class so it probably belongs there rather than as a helper.

The helper and union methods are basically the same so I think only one of them is needed. Probably just something like:
```
@classmethod
def _union_setkind(cls, kinds):
     kinds = set(kinds)
     if not kinds:
         return SetKind()
    elif if len(kinds) == 1:
         return SetKind(kinds.pop())
    else:
         return SetKind(UndefinedKind)
```
The FiniteSet can do
```
kinds = [e.kind for e in self.args]
return self._union_setkind(kinds)
```
And Union can do
```
kinds = [s.kind.element_kind for s in self.args if s is not S.EmptySet]
return self._union_setkind(kinds)
```
And Intersection can do
```
kinds = [s.kind.element_kind for s in self.args if s is not S.UniversalSet]
return self._union_setkind(kinds)
```

Reply:
```python
    def _union_intersection_setkind(self, Union=True):
        if Union:
            Set_type = S.EmptySet
            Kind_1 = SetKind()
            Kind_2 = SetKind(UndefinedKind)
        else:
            Set_type = S.UniversalSet
            Kind_1 = SetKind(UndefinedKind)
            Kind_2 = SetKind()
        args = tuple(arg.kind for arg in cls.args if arg is not Set_type)
        if not args:
            return Kind_1
        elif all(i == args[0] for i in args):
            return args[0]
        else:
            return Kind_2
```
Will this work?

Reply:
I think it's better just to handle the difference between union and intersection outside of this method.

I'm wondering now if a lot of the complication is coming from having EmpySet.kind as SetKind(). Maybe everything would be simpler if the empty set just had `SetKind(UndefinedKind)` but we make sure to exclude it from various situations e.g. `Union` and `Intersection` can handle `EmptySet` specially.

Reply:
> I think it's better just to handle the difference between union and intersection outside of this method.

Will do that

I think keeping EmpySet.kind as SetKind() is good as it differentiates between 2 types of set

Reply:
@oscarbenjamin 
I think keeping EmpySet.kind as SetKind() is good as it differentiates between Universal and EmptySet.
In latest commit I removed property decorator from def _kind